File: sdks/sandbox/csharp/src/OpenSandbox/Models/Mount.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

namespace OpenSandbox.Models;

/// <summary>
/// Selects between ossfs 1.x and ossfs 2.x.
/// </summary>
public enum OssfsVersion
{
    /// <summary>ossfs 1.x</summary>
    Ossfs10,

    /// <summary>ossfs 2.x</summary>
    Ossfs20,
}

/// <summary>
/// Marker base type shared by <see cref="NfsMountOptions"/> and <see cref="OssfsMountOptions"/>.
/// Consumers should use the concrete option types; the base type is only used to
/// keep the extension method signatures narrow.
/// </summary>
public abstract class MountOptions
{
    internal MountOptions()
    {
    }
}

/// <summary>
/// Options describing an NFS mount executed inside a sandbox via
/// <c>sandbox.MountAsync(...)</c>.
/// </summary>
/// <remarks>
/// The SDK does not manage the NAS service. It assembles a <c>mount -t nfs</c>
/// shell command and runs it via <c>sandbox.Commands.RunAsync</c>. The sandbox
/// image must already contain the <c>mount.nfs</c> binary, or
/// <see cref="Installation"/> can install it at mount time.
/// </remarks>
public class NfsMountOptions : MountOptions
{
    /// <summary>
    /// Default NFSv3 option string applied when <see cref="Options"/> is null or blank.
    /// </summary>
    public const string DefaultNfsOptions =
        "vers=3,nolock,proto=tcp,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2,noresvport";

    /// <summary>Gets or sets the NFS server endpoint (host or IP, without the export path).</summary>
    public required string Endpoint { get; set; }

    /// <summary>Gets or sets the export path on the NFS server (for example <c>/</c>).</summary>
    public required string NasPath { get; set; }

    /// <summary>Gets or sets the absolute path inside the sandbox where the export will be mounted.</summary>
    public required string MountPoint { get; set; }

    /// <summary>
    /// Gets or sets the comma-separated NFS mount options passed to <c>mount -o</c>.
    /// Defaults to <see cref="DefaultNfsOptions"/> when null or blank.
    /// </summary>
    public string? Options { get; set; }

    /// <summary>
    /// Gets or sets an optional shell command executed before the mount command,
    /// typically to install NFS client packages (for example <c>apt-get install -y nfs-common</c>).
    /// </summary>
    public string? Installation { get; set; }
}

/// <summary>
/// Options describing an Alibaba Cloud OSS mount executed inside a sandbox with
/// either <c>ossfs</c> (1.x) or <c>ossfs2</c> (2.x).
/// </summary>
/// <remarks>
/// When <see cref="Version"/> is null the SDK defaults to <see cref="OssfsVersion.Ossfs10"/>.
/// <see cref="BucketDirectory"/> is supported by both versions: ossfs 1.x mounts
/// <c>bucket:/dir</c>; ossfs 2.x writes it to the configuration file as
/// <c>--oss_bucket_prefix=&lt;dir&gt;/</c>.
/// </remarks>
public class OssfsMountOptions : MountOptions
{
    /// <summary>Gets or sets the OSS endpoint URL.</summary>
    public required string Endpoint { get; set; }

    /// <summary>Gets or sets the OSS bucket name.</summary>
    public required string Bucket { get; set; }

    /// <summary>Gets or sets the absolute path inside the sandbox where the bucket is mounted.</summary>
    public required string MountPoint { get; set; }

    /// <summary>Gets or sets the Alibaba Cloud access key id.</summary>
    public required string AccessKeyId { get; set; }

    /// <summary>Gets or sets the Alibaba Cloud access key secret.</summary>
    public required string AccessKeySecret { get; set; }

    /// <summary>
    /// Gets or sets the optional STS security token. When set, ossfs 1.x appends
    /// it to the password file and ossfs 2.x exports it as <c>OSS_SESSION_TOKEN</c>.
    /// </summary>
    public string? SecurityToken { get; set; }

    /// <summary>
    /// Gets or sets the ossfs major version. Defaults to <see cref="OssfsVersion.Ossfs10"/> when null.
    /// </summary>
    public OssfsVersion? Version { get; set; }

    /// <summary>Gets or sets an optional subdirectory inside the bucket to mount.</summary>
    public string? BucketDirectory { get; set; }

    /// <summary>
    /// Gets or sets the additional ossfs options. Each entry is the raw option
    /// value without the leading <c>-o</c> (ossfs 1.x) or <c>--</c> (ossfs 2.x)
    /// prefix, for example <c>"use_cache=/tmp/ossfs"</c>.
    /// </summary>
    public IReadOnlyList<string>? Options { get; set; }

    /// <summary>Gets or sets an optional shell command executed before the mount command.</summary>
    public string? Installation { get; set; }
}

File: sdks/sandbox/go/mount.go
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package opensandbox

// This file exposes NAS/OSS mount syntax sugar on top of Sandbox.RunCommand.
// See Sandbox.Mount and Sandbox.Umount below for the public entrypoints and
// mount_options.go for the option types.

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"path/filepath"
	"strings"
)

// DefaultNfsOptions is the option string applied to NfsMountOptions.Options
// when it is empty. It matches the NFSv3 profile used by the Alibaba internal
// SDK for compatibility with NAS servers deployed in production.
const DefaultNfsOptions = "vers=3,nolock,proto=tcp,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2,noresvport"

// ossfs1PasswdPathPrefix is the prefix used to build a per-call ossfs 1.x
// password file under /tmp. A random suffix is appended per mount so that
// concurrent ossfs 1.x mounts in the same sandbox do not overwrite or delete
// each other's credentials. Kept as a package-level constant so tests can
// assert on the exact command emitted by the builders.
const ossfs1PasswdPathPrefix = "/tmp/opensandbox-ossfspass-"

// MountFailedError is returned when a NAS or OSS mount syntax-sugar call fails
// inside the sandbox. The failing Execution is attached so callers can inspect
// exit code, stdout and stderr for diagnostics.
type MountFailedError struct {
	Message   string
	Execution *Execution
}

// Error implements the error interface.
func (e *MountFailedError) Error() string {
	if e == nil {
		return "opensandbox: mount failed"
	}
	return e.Message
}

// Mount runs a NAS or OSS mount command inside this sandbox. The option
// argument selects the backend: pass a *NfsMountOptions for NFS or
// *OssfsMountOptions for OSS (ossfs 1.x / 2.x). Both option types are
// documented in mount_options.go.
//
// The sandbox image must have the corresponding mount binary installed. Use
// the Installation field on the options to install it at mount time.
//
// Returns the Execution from the underlying RunCommand call. If the mount
// command fails (non-zero exit or streamed error event) a *MountFailedError
// is returned; the Execution is attached to the error for diagnostics.
func (s *Sandbox) Mount(ctx context.Context, options MountOptions) (*Execution, error) {
	if s == nil {
		return nil, &InvalidArgumentError{Field: "sandbox", Message: "nil sandbox receiver"}
	}
	switch opt := options.(type) {
	case *NfsMountOptions:
		if opt == nil {
			return nil, &InvalidArgumentError{Field: "options", Message: "nil NfsMountOptions"}
		}
		if err := validateNfs(opt); err != nil {
			return nil, err
		}
		cmd := buildNfsCommand(opt)
		return s.runMount(ctx, cmd, "NAS mount failure")
	case *OssfsMountOptions:
		if opt == nil {
			return nil, &InvalidArgumentError{Field: "options", Message: "nil OssfsMountOptions"}
		}
		if err := validateOssfs(opt); err != nil {
			return nil, err
		}
		switch selectOssfsVersion(opt) {
		case OssfsVersion10:
			return s.runMount(ctx, buildOssfs1Command(opt), "ossfs1.0 mount failure")
		case OssfsVersion20:
			return s.mountOssfs2(ctx, opt)
		default:
			return nil, &InvalidArgumentError{Field: "options.Version", Message: fmt.Sprintf("unsupported ossfs version: %q", opt.Version)}
		}
	default:
		return nil, &InvalidArgumentError{Field: "options", Message: fmt.Sprintf("unsupported mount options type: %T", options)}
	}
}

// Umount unmounts a previously mounted path inside this sandbox by running
// `umount <mount_point>`.
func (s *Sandbox) Umount(ctx context.Context, mountPoint string) (*Execution, error) {
	if s == nil {
		return nil, &InvalidArgumentError{Field: "sandbox", Message: "nil sandbox receiver"}
	}
	cmd, err := buildUmountCommand(mountPoint)
	if err != nil {
		return nil, err
	}
	return s.runMount(ctx, cmd, "umount failure")
}

func (s *Sandbox) mountOssfs2(ctx context.Context, opt *OssfsMountOptions) (*Execution, error) {
	plan := buildOssfs2Plan(opt)
	entry := UploadFileEntry{
		File: strings.NewReader(plan.ConfContent),
		Options: UploadFileOptions{
			FileName: filepath.Base(plan.ConfPath),
			// Mode is serialized as JSON number, then parsed by execd as an octal
			// string (see components/execd/pkg/web/controller/utils.go); pass the
			// literal decimal 600 rather than the Go octal literal 0o600.
			Metadata: FileMetadata{Path: plan.ConfPath, Mode: 600},
		},
	}
	if err := s.UploadFiles(ctx, []UploadFileEntry{entry}); err != nil {
		return nil, fmt.Errorf("opensandbox: upload ossfs2 conf: %w", err)
	}
	return s.runMount(ctx, plan.Command, "ossfs2.0 mount failure")
}

func (s *Sandbox) runMount(ctx context.Context, command string, failurePrefix string) (*Execution, error) {
	execution, err := s.RunCommand(ctx, command, nil)
	if err != nil {
		return execution, err
	}
	if err := ensureMountSuccess(execution, failurePrefix); err != nil {
		return execution, err
	}
	return execution, nil
}

// --- builders (also called directly from tests) ---

// ossfs2Plan describes the configuration file that must be uploaded before the
// ossfs 2.x mount command runs, plus the command itself.
type ossfs2Plan struct {
	ConfPath    string
	ConfContent string
	Command     string
}

func validateNfs(opt *NfsMountOptions) error {
	if strings.TrimSpace(opt.Endpoint) == "" {
		return &InvalidArgumentError{Field: "Endpoint", Message: "must not be blank"}
	}
	if strings.TrimSpace(opt.MountPoint) == "" {
		return &InvalidArgumentError{Field: "MountPoint", Message: "must not be blank"}
	}
	if strings.TrimSpace(opt.NasPath) == "" {
		return &InvalidArgumentError{Field: "NasPath", Message: "must not be blank"}
	}
	return nil
}

func validateOssfs(opt *OssfsMountOptions) error {
	if strings.TrimSpace(opt.Endpoint) == "" {
		return &InvalidArgumentError{Field: "Endpoint", Message: "must not be blank"}
	}
	if strings.TrimSpace(opt.Bucket) == "" {
		return &InvalidArgumentError{Field: "Bucket", Message: "must not be blank"}
	}
	if strings.TrimSpace(opt.MountPoint) == "" {
		return &InvalidArgumentError{Field: "MountPoint", Message: "must not be blank"}
	}
	if strings.TrimSpace(opt.AccessKeyID) == "" {
		return &InvalidArgumentError{Field: "AccessKeyID", Message: "must not be blank"}
	}
	if strings.TrimSpace(opt.AccessKeySecret) == "" {
		return &InvalidArgumentError{Field: "AccessKeySecret", Message: "must not be blank"}
	}
	return nil
}

func buildNfsCommand(opt *NfsMountOptions) string {
	optString := opt.Options
	if strings.TrimSpace(optString) == "" {
		optString = DefaultNfsOptions
	}
	source := opt.Endpoint + ":" + opt.NasPath
	core := fmt.Sprintf(
		"mkdir -p %s && mount -t nfs -o %s %s %s",
		shQuote(opt.MountPoint),
		shQuote(optString),
		shQuote(source),
		shQuote(opt.MountPoint),
	)
	return prependInstallation(opt.Installation, core)
}

func buildOssfs1Command(opt *OssfsMountOptions) string {
	passwd := fmt.Sprintf("%s:%s:%s", opt.Bucket, opt.AccessKeyID, opt.AccessKeySecret)
	if strings.TrimSpace(opt.SecurityToken) != "" {
		// STS mode: bucket:accessKeyID:accessKeySecret:securityToken
		passwd = fmt.Sprintf("%s:%s", passwd, opt.SecurityToken)
	}

	bucketArg := opt.Bucket
	if strings.TrimSpace(opt.BucketDirectory) != "" {
		bucketArg = fmt.Sprintf("%s:/%s", opt.Bucket, opt.BucketDirectory)
	}

	var optionFlags strings.Builder
	for _, o := range opt.Options {
		optionFlags.WriteString(" -o")
		optionFlags.WriteString(shQuote(o))
	}

	// Use a unique per-call password file under /tmp so that concurrent
	// ossfs 1.x mounts do not overwrite or delete each other's credentials.
	// The file is created with mode 0600 in one atomic step (avoids the
	// world-readable window that a naive `echo > ; chmod 600` would produce).
	// Cleanup is guaranteed by capturing the ossfs exit code and running rm
	// unconditionally.
	passwdPath := ossfs1PasswdPathPrefix + randomHex(16)
	quotedPasswdPath := shQuote(passwdPath)
	core := fmt.Sprintf(
		"ossfs --version && "+
			"install -m 600 /dev/null %s && "+
			"printf %%s %s > %s && "+
			"mkdir -p %s && "+
			"( ossfs %s %s -ourl=%s -opasswd_file=%s%s ); "+
			"__rc=$?; rm -f %s; exit $__rc",
		quotedPasswdPath,
		shQuote(passwd),
		quotedPasswdPath,
		shQuote(opt.MountPoint),
		shQuote(bucketArg),
		shQuote(opt.MountPoint),
		shQuote(opt.Endpoint),
		quotedPasswdPath,
		optionFlags.String(),
		quotedPasswdPath,
	)
	return prependInstallation(opt.Installation, core)
}

func buildOssfs2Plan(opt *OssfsMountOptions) ossfs2Plan {
	var conf strings.Builder
	fmt.Fprintf(&conf, "--oss_endpoint=%s\n", opt.Endpoint)
	fmt.Fprintf(&conf, "--oss_bucket=%s\n", opt.Bucket)
	if strings.TrimSpace(opt.BucketDirectory) != "" {
		prefix := strings.TrimRight(opt.BucketDirectory, "/") + "/"
		fmt.Fprintf(&conf, "--oss_bucket_prefix=%s\n", prefix)
	}
	for _, o := range opt.Options {
		fmt.Fprintf(&conf, "--%s\n", o)
	}

	confPath := "/tmp/opensandbox-ossfs-" + randomHex(16) + ".conf"

	stsExport := ""
	if strings.TrimSpace(opt.SecurityToken) != "" {
		stsExport = fmt.Sprintf(" && export OSS_SESSION_TOKEN=%s", shQuote(opt.SecurityToken))
	}

	core := fmt.Sprintf(
		"ossfs2 --version && "+
			"mkdir -p %s && "+
			"export OSS_ACCESS_KEY_ID=%s && "+
			"export OSS_ACCESS_KEY_SECRET=%s"+
			"%s"+
			" && ossfs2 mount %s -c %s",
		shQuote(opt.MountPoint),
		shQuote(opt.AccessKeyID),
		shQuote(opt.AccessKeySecret),
		stsExport,
		shQuote(opt.MountPoint),
		shQuote(confPath),
	)
	return ossfs2Plan{
		ConfPath:    confPath,
		ConfContent: conf.String(),
		Command:     prependInstallation(opt.Installation, core),
	}
}

func buildUmountCommand(mountPoint string) (string, error) {
	if strings.TrimSpace(mountPoint) == "" {
		return "", &InvalidArgumentError{Field: "mountPoint", Message: "must not be blank"}
	}
	return "umount " + shQuote(mountPoint), nil
}

func selectOssfsVersion(opt *OssfsMountOptions) OssfsVersion {
	if opt.Version == "" {
		return OssfsVersion10
	}
	return opt.Version
}

func ensureMountSuccess(execution *Execution, failurePrefix string) error {
	if execution == nil {
		return &MountFailedError{Message: failurePrefix + ": nil execution result"}
	}
	failed := execution.Error != nil || (execution.ExitCode != nil && *execution.ExitCode != 0)
	if !failed {
		return nil
	}
	parts := make([]string, 0, 2)
	if execution.Error != nil {
		parts = append(parts, fmt.Sprintf("[%s] %s", execution.Error.Name, execution.Error.Value))
	}
	if len(execution.Stderr) > 0 {
		var sb strings.Builder
		for i, m := range execution.Stderr {
			if i > 0 {
				sb.WriteByte('\n')
			}
			sb.WriteString(m.Text)
		}
		parts = append(parts, "stderr="+sb.String())
	}
	msg := failurePrefix
	if len(parts) > 0 {
		msg = fmt.Sprintf("%s: %s", failurePrefix, strings.Join(parts, " | "))
	}
	return &MountFailedError{Message: msg, Execution: execution}
}

func prependInstallation(installation string, core string) string {
	if strings.TrimSpace(installation) == "" {
		return core
	}
	return installation + " && " + core
}

// shQuote quotes value for POSIX shell single-quoted context.
// Any embedded single quote is closed, escaped with a backslash, and reopened,
// producing the standard four-character replacement sequence. The result is
// always safe to embed as one argument to an "sh -c" string. Unlike Python's
// shlex.quote, this always adds quotes for consistent behavior across
// languages.
func shQuote(value string) string {
	return "'" + strings.ReplaceAll(value, "'", "'\\''") + "'"
}

func randomHex(bytes int) string {
	buf := make([]byte, bytes)
	if _, err := rand.Read(buf); err != nil {
		// crypto/rand.Read should not fail on any supported platform; fall back
		// to a fixed marker so we still produce a well-formed path in the
		// impossible failure case.
		return "randfailed"
	}
	return hex.EncodeToString(buf)
}
